resources/log.rb

```diff
@@ -16,28 +16,75 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-property :config_dir, String, default: '/etc/syslog-ng/log.d'
-property :cookbook, String
-property :template_source, String
-property :source, [String, Array, Hash], default: []
-property :filter, [String, Array, Hash], default: []
-property :destination, [String, Array], default: []
-property :flags, [String, Array], default: []
-property :parser, [String, Array], default: []
-property :rewrite, [String, Array], default: []
-property :junction, [Hash, Array], default: []
-property :description, String
+include SyslogNg::Cookbook::GeneralHelpers
+
+property :config_dir, String,
+          default: lazy { "#{syslog_ng_config_dir}/log.d" },
+          description: 'Directory to create configuration file in'
+
+property :cookbook, String,
+          default: 'syslog_ng',
+          description: 'Cookbook to source configuration file template from'
+
+property :template, String,
+          default: 'syslog-ng/log.conf.erb',
+          description: 'Template to use to generate the configuration file'
+
+property :owner, String,
+          default: lazy { syslog_ng_user },
+          description: 'Owner of the generated configuration file'
+
+property :group, String,
+          default: lazy { syslog_ng_group },
+          description: 'Group of the generated configuration file'
+
+property :mode, String,
+          default: '0640',
+          description: 'Filemode of the generated configuration file'
+
+property :description, String,
+          description: 'Unparsed description to add to the configuration file'
+
+property :source, [String, Array, Hash],
```

Comment:
```suggestion
property :source, Hash,
```

I would probably just stick with Hash here to keep the interface clearer and easier to document. If everyone expects it to be a Hash, but you can put a String in there, it then suddenly becomes harder to grok.

Reply:
Just to add to this for hash there is a way of doing validation (callbacks) on the keys to ensure people provide the right keys and get a decent error message when the wrong keys are supplied, see: https://github.com/sous-chefs/keepalived/blob/master/resources/http_get.rb#L1

Reply:
The problem with doing that is that it will remove a large portion of functionality, and if anything, it'd be the String and Array that need to be kept. It's actually highlighted that `:destination` should be the same as well and I'd missed the Hash out.

It's like this because Syslog-NG allows you to configure a complete set of config on a `log` statement, so you can have the source/destination/filter/rewrite/etc definition all-in-one on the log resource for a one off usage. If we remove this then you'd need to create separate resources/configurations even when they aren't reused.

I agree then helpers aren't the easiest thing to maintain but we can thank syslog-ng's *charming* config file format for that. When passed a Hash the same helper